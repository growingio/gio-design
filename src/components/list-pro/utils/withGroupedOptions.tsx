import React, { ComponentType } from 'react';
import { get, groupBy } from 'lodash';

type GetDisplayName = (x: ComponentType) => string;
export const getDisplayName: GetDisplayName = (Component) => Component.displayName || Component.name || 'Component';

const getGroupedOptions = (options: any[]) => {
  const hasNotGroup = !options.some((option: any) => {
    if (option?.groupLabel) {
      return true;
    } 
    return false;
  });

  if (hasNotGroup) {
    return options;
  }

  const groupedOptions = groupBy(options, 'groupLabel');
  return Object.keys(groupedOptions).reduce((opts: any[], groupKey: string) => {
    const group = get(groupedOptions, `${groupKey}.0.group`);
    const groupValue = get(groupedOptions, `${groupKey}.0.groupValue`);
    return opts.concat([
      {
        id: groupValue || groupKey,
        value: groupValue || groupKey,
        name: groupKey,
        label: groupKey,
        type: 'groupLabel',
        group,
      },
      ...groupedOptions[groupKey],
    ]);
  }, []);
};

export interface WithGroupedOptionsProps {
  options: any[];
  groups?: any[];
  originalOptions?: any[];
}

const withGroupedOptions = (Component: any, getGroupedFunc = getGroupedOptions): any => {
  class HOC extends React.PureComponent<any> {
    public static displayName = `GroupedOptionsProvider(${getDisplayName(Component)})`;

    public render() {
      const { options } = this.props;
      const groupedOptions = getGroupedOptions(options);
      let groups;
      if (getGroupedFunc === getGroupedOptions) {
        groups = groupedOptions;
      } else {
        groups = getGroupedFunc(options);
      }
      return (
        <Component {...this.props} options={groupedOptions} groups={groups} originalOptions={options} />
      );
    }
  }
  return HOC;
};

export default withGroupedOptions;
