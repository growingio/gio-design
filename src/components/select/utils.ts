import React from 'react';
import { concat } from 'lodash';
import { Option, OptionProps } from './interface';

interface group {
    groupLabel?: string;
    groupValue?: string | number;
}
// ReactNode To Options
export function convertNodeToOption(node: React.ReactElement, group: group): Option {
    const {
        props: { value, children, ...restProps },
    } = node as React.ReactElement & { props: OptionProps };
    const { groupValue, groupLabel } = group;
    return { value, label: children !== undefined ? children : value, groupValue, groupLabel, ...restProps };
}

export function convertChildrenToData(nodes: React.ReactNode, group = {}): Option[] {
    let nodeOptions: Option[] = [];
    React.Children.forEach(nodes, (node: React.ReactElement) => {
        if (!React.isValidElement(node)) {
            return;
        }
        const {
            type: { isSelectOptGroup },
            props: { children, label, value },
        } = node as React.ReactElement & { type: { isSelectOptGroup?: boolean }, props: OptionProps }; // 联合类型
        if (!isSelectOptGroup) {
            // option
            nodeOptions.push(convertNodeToOption(node, group));
        } else {
            // Group
            nodeOptions = concat(nodeOptions, convertChildrenToData(children, { groupLabel: label, groupValue: value }));
        }
    });
    return nodeOptions;
}
