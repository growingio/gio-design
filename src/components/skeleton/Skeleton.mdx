import { Subtitle, ArgsTable } from '@storybook/addon-docs/blocks';
import Skeleton from './Skeleton';

# Skeleton 加载中

<Subtitle>在需要等待加载内容的位置提供一个占位图形组合。</Subtitle>

1. 网络较慢，需要长时间等待加载处理的情况下。
2. 图文信息内容较多的列表/卡片中。
3. 只在第一次加载数据的时候使用。
4. 可以被 loading 完全代替，但是在可用的场景下可以比 loading 提供更好的视觉效果和用户体验。

## 参数说明

  <ArgsTable of={Skeleton} />

## 代码演示

### 骨架图
最简单的占位效果。

[Example](/?path=/story/basic-components-skeleton--default)

### 占位图
图片占位

[Example](/?path=/story/basic-components-skeleton--image)
