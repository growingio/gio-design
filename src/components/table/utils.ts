import { clone, has, get, join, set } from 'lodash';
import { ColumnsType, InnerColumnsType } from './interface';

export const translateInnerColumns = <RecordType>(columns: ColumnsType<RecordType>): InnerColumnsType<RecordType> =>
  clone(columns).map((cloneColumn) => {
    if (!has(cloneColumn, 'key')) {
      if (has(cloneColumn, 'dataIndex')) {
        if (Array.isArray(get(cloneColumn, 'dataIndex'))) {
          set(cloneColumn, 'key', join(get(cloneColumn, 'dataIndex'), '-'));
        } else {
          set(cloneColumn, 'key', get(cloneColumn, 'dataIndex'));
        }
      } else {
        // eslint-disable-next-line no-console
        console.warn('gio-design table: column key or dataIndex must have one');
      }
    }
    if (has(cloneColumn, 'children')) {
      set(cloneColumn, 'children', translateInnerColumns(get(cloneColumn, 'children')));
    }
    return cloneColumn;
  }) as InnerColumnsType<RecordType>;

export default {};
