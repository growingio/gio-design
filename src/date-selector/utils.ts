import format from 'date-fns/format';
import { DATE_FORMAT } from './constant';

export function formatDate(date: Date, formatString?: string) {
  return format(date, formatString ?? DATE_FORMAT);
}

export default {
  formatDate,
};
