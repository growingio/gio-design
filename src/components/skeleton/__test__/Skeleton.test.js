import React from 'react';
import { mount, render } from 'enzyme';
import { waitForComponentToPaint } from '../../../utils/test';
import Skeleton from '../index';

describe('Testing Skeleton', () => {
  it('should be stable', () => {
    const wrapper = render(<Skeleton avatar>children</Skeleton>);
    expect(wrapper).toMatchSnapshot();
  });

  it('should be mount, setProps, unmount with no error', () => {
    expect(() => {
      const wrapper = mount(<Skeleton avatar>children</Skeleton>);
      wrapper.setProps({ delay: 1000 });
      wrapper.setProps({ loading: false });
      wrapper.setProps({ title: '加载中' });
      wrapper.setProps({ avatar: { size: 'small' } });
      wrapper.setProps({ paragraph: { row: 5 } });
      wrapper.setProps({ active: false });
      wrapper.unmount();
    }).not.toThrow();
  });

  test('props delay and loading', (done) => {
    const wrapper = mount(
      <Skeleton avatar loading={false}>
        children
      </Skeleton>
    );
    wrapper.setProps({ delay: 1000, loading: true });
    expect(wrapper.exists('.gio-skeleton')).toBe(false);
    waitForComponentToPaint(wrapper, 1000).then(() => {
      expect(wrapper.exists('.gio-skeleton')).toBe(true);
      done();
    });
  });

  test('props title', () => {
    const wrapper = mount(<Skeleton avatar>children</Skeleton>);
    wrapper.setProps({ title: false });
    expect(wrapper.exists('.gio-skeleton-title')).toBe(false);
  });

  test('props avatar', () => {
    const wrapper = mount(<Skeleton>children</Skeleton>);
    wrapper.setProps({ avatar: { size: 'small' } });
    expect(wrapper.exists('.gio-skeleton-header')).toBe(true);
    expect(wrapper.exists('.gio-skeleton-content')).toBe(true);
  });

  test('props paragraph', () => {
    const wrapper = mount(<Skeleton>children</Skeleton>);
    wrapper.setProps({ paragraph: { row: 5 } });
    expect(wrapper.find('.gio-skeleton-paragraph').children()).toHaveLength(5);
    wrapper.setProps({ paragraph: false, title: false });
    expect(wrapper.exists('.gio-skeleton-paragraph')).toBe(false);
    expect(wrapper.exists('.gio-skeleton-title')).toBe(false);
    expect(wrapper.exists('.gio-skeleton-content')).toBe(false);
  });

  test('props active', () => {
    const wrapper = mount(<Skeleton active>children</Skeleton>);
    expect(wrapper.exists('.gio-skeleton-active')).toBe(true);
  });

  test('Skeleton Image', (done) => {
    const wrapper = mount(<Skeleton.Image loading>children</Skeleton.Image>);
    expect(wrapper.exists('.gio-skeleton-image')).toBe(true);
    wrapper.setProps({ loading: false });
    waitForComponentToPaint(wrapper, 1000).then(() => {
      expect(wrapper.exists('.gio-skeleton-image')).toBe(false);
      done();
    });
  });
});
