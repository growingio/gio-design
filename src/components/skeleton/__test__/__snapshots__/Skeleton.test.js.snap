// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Skeleton should be stable 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "gio-skeleton gio-skeleton-active",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": "gio-skeleton-header",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "class": "gio-skeleton-avatar gio-avatar gio-avatar-lg",
            },
            "children": Array [
              Object {
                "attribs": Object {
                  "class": "gio-avatar-default",
                },
                "children": Array [
                  Object {
                    "attribs": Object {
                      "class": "gio-icon",
                    },
                    "children": Array [
                      Object {
                        "attribs": Object {
                          "class": "gio-icon-svg",
                          "fill": "currentColor",
                          "height": "24",
                          "style": "color:#fff",
                          "viewBox": "0 0 14 14",
                          "width": "24",
                          "xmlns": "http://www.w3.org/2000/svg",
                        },
                        "children": Array [
                          Object {
                            "attribs": Object {
                              "clip-rule": "evenodd",
                              "d": "M13 13.6s0-.1 0 0C12.8 10.4 10.2 8 7 8s-5.8 2.4-6 5.5c0 .3-.2.5-.5.5s-.5-.2-.5-.5c.2-2.6 1.8-4.8 4.1-5.8l.3-.3c0-.2 0-.3-.1-.4C3.5 6.1 3 5.1 3 4c0-2.2 1.8-4 4-4s4 1.8 4 4c0 1.1-.5 2.1-1.2 2.9-.1.1-.2.2-.2.4s.1.3.3.3c2.3 1 3.9 3.2 4.1 5.8v.1c0 .3-.2.5-.5.5s-.5-.2-.5-.4zM7 1c1.7 0 3 1.3 3 3S8.7 7 7 7 4 5.7 4 4s1.3-3 3-3z",
                              "fill": "currentColor",
                              "fill-rule": "evenodd",
                            },
                            "children": Array [],
                            "name": "path",
                            "namespace": "http://www.w3.org/2000/svg",
                            "next": null,
                            "parent": [Circular],
                            "prev": null,
                            "type": "tag",
                            "x-attribsNamespace": Object {
                              "clip-rule": undefined,
                              "d": undefined,
                              "fill": undefined,
                              "fill-rule": undefined,
                            },
                            "x-attribsPrefix": Object {
                              "clip-rule": undefined,
                              "d": undefined,
                              "fill": undefined,
                              "fill-rule": undefined,
                            },
                          },
                        ],
                        "name": "svg",
                        "namespace": "http://www.w3.org/2000/svg",
                        "next": null,
                        "parent": [Circular],
                        "prev": null,
                        "type": "tag",
                        "x-attribsNamespace": Object {
                          "class": undefined,
                          "fill": undefined,
                          "height": undefined,
                          "style": undefined,
                          "viewBox": undefined,
                          "width": undefined,
                          "xmlns": "http://www.w3.org/2000/xmlns/",
                        },
                        "x-attribsPrefix": Object {
                          "class": undefined,
                          "fill": undefined,
                          "height": undefined,
                          "style": undefined,
                          "viewBox": undefined,
                          "width": undefined,
                          "xmlns": "",
                        },
                      },
                    ],
                    "name": "span",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "tag",
                    "x-attribsNamespace": Object {
                      "class": undefined,
                    },
                    "x-attribsPrefix": Object {
                      "class": undefined,
                    },
                  },
                ],
                "name": "span",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                },
              },
            ],
            "name": "span",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "class": "gio-skeleton-content",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "class": "gio-skeleton-title",
              },
              "children": Array [],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": Object {
                "attribs": Object {
                  "class": "gio-skeleton-paragraph",
                },
                "children": Array [
                  Object {
                    "attribs": Object {},
                    "children": Array [],
                    "name": "p",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": Object {
                      "attribs": Object {},
                      "children": Array [],
                      "name": "p",
                      "namespace": "http://www.w3.org/1999/xhtml",
                      "next": Object {
                        "attribs": Object {},
                        "children": Array [],
                        "name": "p",
                        "namespace": "http://www.w3.org/1999/xhtml",
                        "next": null,
                        "parent": [Circular],
                        "prev": [Circular],
                        "type": "tag",
                        "x-attribsNamespace": Object {},
                        "x-attribsPrefix": Object {},
                      },
                      "parent": [Circular],
                      "prev": [Circular],
                      "type": "tag",
                      "x-attribsNamespace": Object {},
                      "x-attribsPrefix": Object {},
                    },
                    "parent": [Circular],
                    "prev": null,
                    "type": "tag",
                    "x-attribsNamespace": Object {},
                    "x-attribsPrefix": Object {},
                  },
                  Object {
                    "attribs": Object {},
                    "children": Array [],
                    "name": "p",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": Object {
                      "attribs": Object {},
                      "children": Array [],
                      "name": "p",
                      "namespace": "http://www.w3.org/1999/xhtml",
                      "next": null,
                      "parent": [Circular],
                      "prev": [Circular],
                      "type": "tag",
                      "x-attribsNamespace": Object {},
                      "x-attribsPrefix": Object {},
                    },
                    "parent": [Circular],
                    "prev": Object {
                      "attribs": Object {},
                      "children": Array [],
                      "name": "p",
                      "namespace": "http://www.w3.org/1999/xhtml",
                      "next": [Circular],
                      "parent": [Circular],
                      "prev": null,
                      "type": "tag",
                      "x-attribsNamespace": Object {},
                      "x-attribsPrefix": Object {},
                    },
                    "type": "tag",
                    "x-attribsNamespace": Object {},
                    "x-attribsPrefix": Object {},
                  },
                  Object {
                    "attribs": Object {},
                    "children": Array [],
                    "name": "p",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": null,
                    "parent": [Circular],
                    "prev": Object {
                      "attribs": Object {},
                      "children": Array [],
                      "name": "p",
                      "namespace": "http://www.w3.org/1999/xhtml",
                      "next": [Circular],
                      "parent": [Circular],
                      "prev": Object {
                        "attribs": Object {},
                        "children": Array [],
                        "name": "p",
                        "namespace": "http://www.w3.org/1999/xhtml",
                        "next": [Circular],
                        "parent": [Circular],
                        "prev": null,
                        "type": "tag",
                        "x-attribsNamespace": Object {},
                        "x-attribsPrefix": Object {},
                      },
                      "type": "tag",
                      "x-attribsNamespace": Object {},
                      "x-attribsPrefix": Object {},
                    },
                    "type": "tag",
                    "x-attribsNamespace": Object {},
                    "x-attribsPrefix": Object {},
                  },
                ],
                "name": "div",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": [Circular],
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                },
              },
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
            Object {
              "attribs": Object {
                "class": "gio-skeleton-paragraph",
              },
              "children": Array [
                Object {
                  "attribs": Object {},
                  "children": Array [],
                  "name": "p",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": Object {
                    "attribs": Object {},
                    "children": Array [],
                    "name": "p",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": Object {
                      "attribs": Object {},
                      "children": Array [],
                      "name": "p",
                      "namespace": "http://www.w3.org/1999/xhtml",
                      "next": null,
                      "parent": [Circular],
                      "prev": [Circular],
                      "type": "tag",
                      "x-attribsNamespace": Object {},
                      "x-attribsPrefix": Object {},
                    },
                    "parent": [Circular],
                    "prev": [Circular],
                    "type": "tag",
                    "x-attribsNamespace": Object {},
                    "x-attribsPrefix": Object {},
                  },
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {},
                  "x-attribsPrefix": Object {},
                },
                Object {
                  "attribs": Object {},
                  "children": Array [],
                  "name": "p",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": Object {
                    "attribs": Object {},
                    "children": Array [],
                    "name": "p",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": null,
                    "parent": [Circular],
                    "prev": [Circular],
                    "type": "tag",
                    "x-attribsNamespace": Object {},
                    "x-attribsPrefix": Object {},
                  },
                  "parent": [Circular],
                  "prev": Object {
                    "attribs": Object {},
                    "children": Array [],
                    "name": "p",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": [Circular],
                    "parent": [Circular],
                    "prev": null,
                    "type": "tag",
                    "x-attribsNamespace": Object {},
                    "x-attribsPrefix": Object {},
                  },
                  "type": "tag",
                  "x-attribsNamespace": Object {},
                  "x-attribsPrefix": Object {},
                },
                Object {
                  "attribs": Object {},
                  "children": Array [],
                  "name": "p",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": Object {
                    "attribs": Object {},
                    "children": Array [],
                    "name": "p",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": [Circular],
                    "parent": [Circular],
                    "prev": Object {
                      "attribs": Object {},
                      "children": Array [],
                      "name": "p",
                      "namespace": "http://www.w3.org/1999/xhtml",
                      "next": [Circular],
                      "parent": [Circular],
                      "prev": null,
                      "type": "tag",
                      "x-attribsNamespace": Object {},
                      "x-attribsPrefix": Object {},
                    },
                    "type": "tag",
                    "x-attribsNamespace": Object {},
                    "x-attribsPrefix": Object {},
                  },
                  "type": "tag",
                  "x-attribsNamespace": Object {},
                  "x-attribsPrefix": Object {},
                },
              ],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": Object {
                "attribs": Object {
                  "class": "gio-skeleton-title",
                },
                "children": Array [],
                "name": "div",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": [Circular],
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                },
              },
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
      Object {
        "attribs": Object {
          "class": "gio-skeleton-content",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "class": "gio-skeleton-title",
            },
            "children": Array [],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": Object {
              "attribs": Object {
                "class": "gio-skeleton-paragraph",
              },
              "children": Array [
                Object {
                  "attribs": Object {},
                  "children": Array [],
                  "name": "p",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": Object {
                    "attribs": Object {},
                    "children": Array [],
                    "name": "p",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": Object {
                      "attribs": Object {},
                      "children": Array [],
                      "name": "p",
                      "namespace": "http://www.w3.org/1999/xhtml",
                      "next": null,
                      "parent": [Circular],
                      "prev": [Circular],
                      "type": "tag",
                      "x-attribsNamespace": Object {},
                      "x-attribsPrefix": Object {},
                    },
                    "parent": [Circular],
                    "prev": [Circular],
                    "type": "tag",
                    "x-attribsNamespace": Object {},
                    "x-attribsPrefix": Object {},
                  },
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {},
                  "x-attribsPrefix": Object {},
                },
                Object {
                  "attribs": Object {},
                  "children": Array [],
                  "name": "p",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": Object {
                    "attribs": Object {},
                    "children": Array [],
                    "name": "p",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": null,
                    "parent": [Circular],
                    "prev": [Circular],
                    "type": "tag",
                    "x-attribsNamespace": Object {},
                    "x-attribsPrefix": Object {},
                  },
                  "parent": [Circular],
                  "prev": Object {
                    "attribs": Object {},
                    "children": Array [],
                    "name": "p",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": [Circular],
                    "parent": [Circular],
                    "prev": null,
                    "type": "tag",
                    "x-attribsNamespace": Object {},
                    "x-attribsPrefix": Object {},
                  },
                  "type": "tag",
                  "x-attribsNamespace": Object {},
                  "x-attribsPrefix": Object {},
                },
                Object {
                  "attribs": Object {},
                  "children": Array [],
                  "name": "p",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": Object {
                    "attribs": Object {},
                    "children": Array [],
                    "name": "p",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": [Circular],
                    "parent": [Circular],
                    "prev": Object {
                      "attribs": Object {},
                      "children": Array [],
                      "name": "p",
                      "namespace": "http://www.w3.org/1999/xhtml",
                      "next": [Circular],
                      "parent": [Circular],
                      "prev": null,
                      "type": "tag",
                      "x-attribsNamespace": Object {},
                      "x-attribsPrefix": Object {},
                    },
                    "type": "tag",
                    "x-attribsNamespace": Object {},
                    "x-attribsPrefix": Object {},
                  },
                  "type": "tag",
                  "x-attribsNamespace": Object {},
                  "x-attribsPrefix": Object {},
                },
              ],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": [Circular],
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
          Object {
            "attribs": Object {
              "class": "gio-skeleton-paragraph",
            },
            "children": Array [
              Object {
                "attribs": Object {},
                "children": Array [],
                "name": "p",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": Object {
                  "attribs": Object {},
                  "children": Array [],
                  "name": "p",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": Object {
                    "attribs": Object {},
                    "children": Array [],
                    "name": "p",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": null,
                    "parent": [Circular],
                    "prev": [Circular],
                    "type": "tag",
                    "x-attribsNamespace": Object {},
                    "x-attribsPrefix": Object {},
                  },
                  "parent": [Circular],
                  "prev": [Circular],
                  "type": "tag",
                  "x-attribsNamespace": Object {},
                  "x-attribsPrefix": Object {},
                },
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
              Object {
                "attribs": Object {},
                "children": Array [],
                "name": "p",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": Object {
                  "attribs": Object {},
                  "children": Array [],
                  "name": "p",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": [Circular],
                  "type": "tag",
                  "x-attribsNamespace": Object {},
                  "x-attribsPrefix": Object {},
                },
                "parent": [Circular],
                "prev": Object {
                  "attribs": Object {},
                  "children": Array [],
                  "name": "p",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": [Circular],
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {},
                  "x-attribsPrefix": Object {},
                },
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
              Object {
                "attribs": Object {},
                "children": Array [],
                "name": "p",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": Object {
                  "attribs": Object {},
                  "children": Array [],
                  "name": "p",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": [Circular],
                  "parent": [Circular],
                  "prev": Object {
                    "attribs": Object {},
                    "children": Array [],
                    "name": "p",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": [Circular],
                    "parent": [Circular],
                    "prev": null,
                    "type": "tag",
                    "x-attribsNamespace": Object {},
                    "x-attribsPrefix": Object {},
                  },
                  "type": "tag",
                  "x-attribsNamespace": Object {},
                  "x-attribsPrefix": Object {},
                },
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": Object {
              "attribs": Object {
                "class": "gio-skeleton-title",
              },
              "children": Array [],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": [Circular],
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "class": "gio-skeleton-header",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "class": "gio-skeleton-avatar gio-avatar gio-avatar-lg",
              },
              "children": Array [
                Object {
                  "attribs": Object {
                    "class": "gio-avatar-default",
                  },
                  "children": Array [
                    Object {
                      "attribs": Object {
                        "class": "gio-icon",
                      },
                      "children": Array [
                        Object {
                          "attribs": Object {
                            "class": "gio-icon-svg",
                            "fill": "currentColor",
                            "height": "24",
                            "style": "color:#fff",
                            "viewBox": "0 0 14 14",
                            "width": "24",
                            "xmlns": "http://www.w3.org/2000/svg",
                          },
                          "children": Array [
                            Object {
                              "attribs": Object {
                                "clip-rule": "evenodd",
                                "d": "M13 13.6s0-.1 0 0C12.8 10.4 10.2 8 7 8s-5.8 2.4-6 5.5c0 .3-.2.5-.5.5s-.5-.2-.5-.5c.2-2.6 1.8-4.8 4.1-5.8l.3-.3c0-.2 0-.3-.1-.4C3.5 6.1 3 5.1 3 4c0-2.2 1.8-4 4-4s4 1.8 4 4c0 1.1-.5 2.1-1.2 2.9-.1.1-.2.2-.2.4s.1.3.3.3c2.3 1 3.9 3.2 4.1 5.8v.1c0 .3-.2.5-.5.5s-.5-.2-.5-.4zM7 1c1.7 0 3 1.3 3 3S8.7 7 7 7 4 5.7 4 4s1.3-3 3-3z",
                                "fill": "currentColor",
                                "fill-rule": "evenodd",
                              },
                              "children": Array [],
                              "name": "path",
                              "namespace": "http://www.w3.org/2000/svg",
                              "next": null,
                              "parent": [Circular],
                              "prev": null,
                              "type": "tag",
                              "x-attribsNamespace": Object {
                                "clip-rule": undefined,
                                "d": undefined,
                                "fill": undefined,
                                "fill-rule": undefined,
                              },
                              "x-attribsPrefix": Object {
                                "clip-rule": undefined,
                                "d": undefined,
                                "fill": undefined,
                                "fill-rule": undefined,
                              },
                            },
                          ],
                          "name": "svg",
                          "namespace": "http://www.w3.org/2000/svg",
                          "next": null,
                          "parent": [Circular],
                          "prev": null,
                          "type": "tag",
                          "x-attribsNamespace": Object {
                            "class": undefined,
                            "fill": undefined,
                            "height": undefined,
                            "style": undefined,
                            "viewBox": undefined,
                            "width": undefined,
                            "xmlns": "http://www.w3.org/2000/xmlns/",
                          },
                          "x-attribsPrefix": Object {
                            "class": undefined,
                            "fill": undefined,
                            "height": undefined,
                            "style": undefined,
                            "viewBox": undefined,
                            "width": undefined,
                            "xmlns": "",
                          },
                        },
                      ],
                      "name": "span",
                      "namespace": "http://www.w3.org/1999/xhtml",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "tag",
                      "x-attribsNamespace": Object {
                        "class": undefined,
                      },
                      "x-attribsPrefix": Object {
                        "class": undefined,
                      },
                    },
                  ],
                  "name": "span",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "class": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "class": undefined,
                  },
                },
              ],
              "name": "span",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
